include/engine/phantom_node.hpp

```diff
@@ -134,6 +105,23 @@ struct PhantomNode
 
     bool IsValid() const { return location.IsValid() && (name_id != INVALID_NAMEID); }
 
+    bool IsForwardValidSource() const
```

Comment:
The naming is a little bit weird. In English `IsValidForwardSource` would make more sense, same for the other functions.